shotover/src/transforms/query_counter.rs

```diff
@@ -26,11 +29,25 @@ pub struct QueryCounterConfig {
 
 impl QueryCounter {
     pub fn new(counter_name: String) -> Self {
-        counter!("shotover_query_count", "name" => counter_name.clone());
+        let counter_name_ref: &'static str = counter_name.leak();
+        let mut counters = HashMap::new();
+        let query_counter = counter!("shotover_query_count", "name" => counter_name_ref);
+        counters.insert("shotover_query_count".to_string(), query_counter);
 
         QueryCounter {
             // Leaking here is fine since the builder is created only once during shotover startup.
-            counter_name: counter_name.leak(),
+            counter_name: counter_name_ref,
+            counters,
+        }
+    }
+
+    fn increment_counter(&mut self, query: String, query_type: &'static str) {
```

Comment:
To efficiently handle both owned and reference cases we can split this into:
```rust
fn increment_counter(&mut self, query: String, query_type: &'static str)
fn increment_counter_ref(&mut self, query: &str, query_type: &'static str)
```
This way if we are using the cassandra protocol we can completely avoid cloning the string when fetching from the hashmap since query is an `&str`.
While at the same time we can avoid an unncessesary clone for redis where query is a `String`.

## Alternatively
Or we could use [Cow](https://doc.rust-lang.org/std/borrow/enum.Cow.html) type to handle both cases from the one function:
```rust
fn increment_counter(&mut self, query: Cow<str>, query_type: &'static str) { .. }

self.increment_counter(Cow::borrowed("foo"), "kafka")
self.increment_counter(Cow::owned("foo".to_string()), "kafka")
```

The cow solution is probably nicer, but duplicating the function is probably easier.
I would happily accept either solution.

Reply:
I'll try to use Cow